processing/src/main/java/org/apache/druid/segment/nested/VariantColumnSerializer.java

```diff
@@ -299,155 +301,241 @@ public void serialize(ColumnValueSelector<? extends StructuredData> selector) th
+    private final boolean writeDictionary;
+
+    private final DictionaryIdLookup dictionaryIdLookup;
+
+    public VariantColumnSerializerizer(
+        String columnName,
+        Byte variantTypeSetByte,
+        DictionaryWriter<String> dictionaryWriter,
+        FixedIndexedWriter<Long> longDictionaryWriter,
+        FixedIndexedWriter<Double> doubleDictionaryWriter,
+        FrontCodedIntArrayIndexedWriter arrayDictionaryWriter,
+        SingleValueColumnarIntsSerializer encodedValueSerializer,
+        GenericIndexedWriter<ImmutableBitmap> bitmapIndexWriter,
+        FixedIndexedIntWriter arrayElementDictionaryWriter,
+        GenericIndexedWriter<ImmutableBitmap> arrayElementIndexWriter,
+        DictionaryIdLookup dictionaryIdLookup,
+        boolean writeDictionary
+    )
+    {
+      this.columnName = columnName;
+      this.columnNameBytes = ColumnSerializerUtils.stringToUtf8InVSizeByteBuffer(columnName);
+      this.variantTypeSetByte = variantTypeSetByte;
+      this.dictionaryWriter = dictionaryWriter;
+      this.longDictionaryWriter = longDictionaryWriter;
+      this.doubleDictionaryWriter = doubleDictionaryWriter;
+      this.arrayDictionaryWriter = arrayDictionaryWriter;
+      this.encodedValueSerializer = encodedValueSerializer;
+      this.bitmapIndexWriter = bitmapIndexWriter;
+      this.arrayElementDictionaryWriter = arrayElementDictionaryWriter;
+      this.arrayElementIndexWriter = arrayElementIndexWriter;
+      this.writeDictionary = writeDictionary;
+      this.dictionaryIdLookup = dictionaryIdLookup;
+
+      boolean[] dictionariesSorted = new boolean[]{
+          dictionaryWriter.isSorted(),
+          longDictionaryWriter.isSorted(),
+          doubleDictionaryWriter.isSorted(),
+          arrayDictionaryWriter.isSorted()
+      };
+      for (boolean sorted : dictionariesSorted) {
+        if (writeDictionary && !sorted) {
+          throw DruidException.defensive(
+              "Dictionary is not sorted? [%s]  Should always be sorted",
+              Arrays.toString(dictionariesSorted)
+          );
         }
       }
-      rowCount++;
     }
 
-    for (int i = 0; i < bitmaps.length; i++) {
-      final MutableBitmap bitmap = bitmaps[i];
-      bitmapIndexWriter.write(
-          indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeImmutableBitmap(bitmap)
-      );
-      bitmaps[i] = null; // Reclaim memory
-    }
-    if (writeDictionary) {
-      for (Int2ObjectMap.Entry<MutableBitmap> arrayElement : arrayElements.int2ObjectEntrySet()) {
-        arrayElementDictionaryWriter.write(arrayElement.getIntKey());
-        arrayElementIndexWriter.write(
-            indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeImmutableBitmap(arrayElement.getValue())
-        );
+    @Override
+    public long getSerializedSize()
+    {
+      long size = 1 + columnNameBytes.capacity();
+      // the value dictionaries, raw column, and null index are all stored in separate files
```

Comment:
nit: i know this isn't new, but this comment is kind of stale and copied from the nested column serializer. the variant column doesn't have a raw column, and instead of 'null index' like the top level nested column has, it should mention that "value dictionaries and indexes and array element indexes, and the dictionary id column are stored in separate files" or something similar

Reply:
Changed